diff --git a/projects/configs/petr/petr_vovnet_gridmask_p4_800x320.py b/projects/configs/petr/petr_vovnet_gridmask_p4_800x320.py
index f6a9614..1e38805 100644
--- a/projects/configs/petr/petr_vovnet_gridmask_p4_800x320.py
+++ b/projects/configs/petr/petr_vovnet_gridmask_p4_800x320.py
@@ -204,8 +204,8 @@ data = dict(
         # we use box_type_3d='LiDAR' in kitti and nuscenes dataset
         # and box_type_3d='Depth' in sunrgbd and scannet dataset.
         box_type_3d='LiDAR'),
-    val=dict(type=dataset_type, pipeline=test_pipeline, classes=class_names, modality=input_modality),
-    test=dict(type=dataset_type, pipeline=test_pipeline, classes=class_names, modality=input_modality))
+    val=dict(type=dataset_type, data_root=data_root, ann_file=data_root + 'nuscenes_infos_val.pkl', pipeline=test_pipeline, classes=class_names, modality=input_modality),
+    test=dict(type=dataset_type, data_root=data_root, ann_file=data_root + 'nuscenes_infos_val.pkl', pipeline=test_pipeline, classes=class_names, modality=input_modality))
 
 optimizer = dict(
     type='AdamW', 
diff --git a/projects/configs/petrv2/petrv2_vovnet_gridmask_p4_800x320.py b/projects/configs/petrv2/petrv2_vovnet_gridmask_p4_800x320.py
index 9b908bc..01a0cc8 100644
--- a/projects/configs/petrv2/petrv2_vovnet_gridmask_p4_800x320.py
+++ b/projects/configs/petrv2/petrv2_vovnet_gridmask_p4_800x320.py
@@ -216,8 +216,8 @@ data = dict(
         # we use box_type_3d='LiDAR' in kitti and nuscenes dataset
         # and box_type_3d='Depth' in sunrgbd and scannet dataset.
         box_type_3d='LiDAR'),
-    val=dict(type=dataset_type, pipeline=test_pipeline, ann_file=data_root + 'mmdet3d_nuscenes_30f_infos_val.pkl', classes=class_names, modality=input_modality),
-    test=dict(type=dataset_type, pipeline=test_pipeline, ann_file=data_root + 'mmdet3d_nuscenes_30f_infos_val.pkl', classes=class_names, modality=input_modality))
+    val=dict(type=dataset_type, pipeline=test_pipeline, data_root=data_root, ann_file=data_root + 'mmdet3d_nuscenes_30f_infos_val.pkl', classes=class_names, modality=input_modality),
+    test=dict(type=dataset_type, pipeline=test_pipeline, data_root=data_root, ann_file=data_root + 'mmdet3d_nuscenes_30f_infos_val.pkl', classes=class_names, modality=input_modality))
 
 
 optimizer = dict(
